// testcafe/lib/services/utils/ipc/packet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Packet = void 0;
const runtime_1 = require("../../../errors/runtime");
const types_1 = require("../../../errors/types");
const HEAD_MASK = 0x01;
const TAIL_MASK = 0x02;
const BYTE_SHIFT = 8;
const BYTE_MASK = 2 ** BYTE_SHIFT - 1;
let Packet = /** @class */ (() => {
    class Packet {
        static _parseHeader(buffer) {
            const dataSize = buffer[1] << BYTE_SHIFT << BYTE_SHIFT | buffer[2] << BYTE_SHIFT | buffer[3];
            return {
                head: Boolean(buffer[0] & HEAD_MASK),
                tail: Boolean(buffer[0] & TAIL_MASK),
                size: dataSize,
                totalSize: dataSize + Packet.HEADER_SIZE
            };
        }
        static parse(buffer) {
            if (buffer.length < Packet.HEADER_SIZE)
                return void 0;
            const header = Packet._parseHeader(buffer);
            if (header.size > this.MAX_PAYLOAD_SIZE)
                throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.tooLargeIPCPayload);
            if (buffer.length < header.size)
                return void 0;
            return { header, data: buffer.slice(Packet.HEADER_SIZE, Packet.HEADER_SIZE + header.size) };
        }
        static _serializeHeader({ size, head, tail }, buffer) {
            buffer[0] = 0;
            if (head)
                buffer[0] |= HEAD_MASK;
            if (tail)
                buffer[0] |= TAIL_MASK;
            buffer[1] = size >> BYTE_SHIFT >> BYTE_SHIFT & BYTE_MASK;
            buffer[2] = size >> BYTE_SHIFT & BYTE_MASK;
            buffer[3] = size & BYTE_MASK;
        }
        static serialize(data, { head = false, tail = false } = {}) {
            const size = data.length;
            if (size > Packet.MAX_PAYLOAD_SIZE)
                throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.tooLargeIPCPayload);
            const buffer = Buffer.alloc(size + Packet.HEADER_SIZE);
            Packet._serializeHeader({ size, head, tail }, buffer);
            data.copy(buffer, Packet.HEADER_SIZE);
            return buffer;
        }
    }
    // NOTE: Max message size: 64 KiB, header size: 4 B
    Packet.MAX_PACKET_SIZE = 64 * 2 ** 10;
    Packet.HEADER_SIZE = 4;
    Packet.MAX_PAYLOAD_SIZE = Packet.MAX_PACKET_SIZE - Packet.HEADER_SIZE;
    return Packet;
})();
exports.Packet = Packet;
exports.default = Packet;
module.exports = exports.default;
